src/julia-parser.scm

```diff
@@ -1254,22 +1262,50 @@
                      (body  (parse-block s)))
                 (expect-end s word)
                 (list word def body)))))
+
        ((abstract)
-        (list 'abstract (parse-subtype-spec s)))
-       ((type immutable)
-        (let ((immu? (eq? word 'immutable)))
-          (if (reserved-word? (peek-token s))
-              (error (string "invalid type name \"" (take-token s) "\"")))
-          (let ((sig (parse-subtype-spec s)))
-            (begin0 (list 'type (if (eq? word 'type) 'true 'false)
-                          sig (parse-block s))
-                    (expect-end s word)))))
+        (let ((ty (eq? (peek-token s) 'type)))
+          (if ty
+              (take-token s))
+          (let ((spec (parse-subtype-spec s)))
+            (if (not ty)
+                (syntax-deprecation s (string "abstract " (deparse spec))
+                                    (string "abstract type " (deparse spec) " end")))
+            (begin0 (list 'abstract spec)
+                    (if ty (expect-end s "abstract type"))))))
+       ((struct)
+        (begin (take-token s)
+               (parse-struct-def s #f word)))
+       ((mutable)
+        (if (not (eq? (peek-token s) 'struct))
+            (parse-call-chain s word #f)
+            (begin (take-token s)
+                   (parse-struct-def s #t word))))
+       ((primitive)
+        (if (not (eq? (peek-token s) 'type))
+            (parse-call-chain s word #f)
+            (begin (take-token s)
+                   (let* ((spec (with-space-sensitive (parse-subtype-spec s)))
+                          (nb   (with-space-sensitive (parse-cond s))))
+                     (begin0 (list 'bitstype nb spec)
+                             (expect-end s "primitive type"))))))
+       ;; deprecated type keywords
+       ((type)
+        (syntax-deprecation s "type" "mutable struct")
+        (parse-struct-def s #t word))
+       ((immutable)
+        (syntax-deprecation s "immutable" "struct")
+        (parse-struct-def s #f word))
```

Comment:
The hurry is that the next release is likely to be 1.0, so removing deprecations can't be done after that until 2.0. We are putting more deprecation churn into 0.6 than we've usually done though.

An alternative that has been suggested on a few calls would be to simultaneously release 0.7 and 1.0 with almost the exact same code, but with any new deprecations done as usual warnings in 0.7 and errors/deleted in 1.0.